chapter_09_commands.asciidoc

```diff
@@ -220,3 +204,150 @@ TODO (ej) Devil's advocate:  If your messagebus.handle processes half the events
+something, we represent their request as a _Command_. That command should modify
+a single _Aggregate_ and either succeed or fail in totality. Any other book
+keeping, clean up, and notification we need to do can happen via an _Event_. We
+don't require the event handlers to succeed in order for the command to be
+successful.
+
+Let's take another example to see why not.
+
+Imagine we are building an e-commerce website that sells expensive luxury goods.
+Our marketing department want to reward customers for repeat visits. We will
+flag customers as VIPs once they make their third purchase, and this will
+entitle them to priority treatment and special offers. Our acceptance criteria
+for this story read as follows:
+
+Given a customer with two orders in their history,
+when the customer places a third order,
+they should be flagged as a VIP.
+
+When a customer first becomes a VIP
+we should send them an email to congratulate them
+
+Using the techniques we've already discussed in this book, we decide that we
+want to build a new History aggregate that records orders and can raise domain
+events when rules are met. We will structure the code like this:
+
+[[new_messagebus]]
+.VIP Customer
+====
+[source,python]
+----
+
+class History (Aggregate):
+
+    def __init__(self, customer_id: int):
+        self.orders = Set() # Set[HistoryEntry]
+        self.customer_id = customer_id
+
+    def record_order(self, order_id: str, order_amount: int): #<1>
+        entry = HistoryEntry(order_id, order_amount)
+
+        if entry in self.orders:
+            return
+
+        self.orders.add(entry)
+
+        if len(self.orders) == 3:
+            self.events.append(
+                CustomerBecameVIP(self.customer_id)
+            )
+
+
+def create_order_from_basket(uow, cmd: CreateOrder): #<2>
+    with uow:
+        order = Order.from_basket(cmd.customer_id, cmd.basket)
+        uow.orders.add(
+        uow.commit() # raises OrderCreated
+
+
+def update_customer_history(uow, event: OrderCreated): #<3>
+    with uow:
+        history = uow.order_history.get(event.customer_id)
+        history.record_order(event.order_id, event.order_amount)
+        uow.commit() # raises CustomerBecameVIP
+
+
+def congratulate_vip_customer(uow, event: CustomerBecameVip): #<4>
+    with uow:
+        customer = uow.customers.get(event.customer_id)
+        email.send(
+            customer.email_address,
+            f'Congratulations {customer.first_name}!'
+        )
+        
+----
+====
+
+<1> The History aggregate captures the rules for when a customer becomes a VIP.
+<2> Our first handler creates an order for the customer and raises a domain
+    event 'OrderCreated'.
+<3> Our second handler updates the History object to record that an order was
+    created.
+<4> Finally we send an email to the customer when they become a VIP using the
+    same email system we used for the OutOfStock event.
+
+//TODO: Sequence diagram here?
+
```

Comment:
Looking forward to seeing the rest as it comes along.  I've not yet caught up to all the updates since July, but here's my 2 cents:

* The idea of "at the same time" implies some kind of transaction and strong consistency.  I think that it can act as a mental block towards reasoning about eventual consistency, since there is never actually such a thing as "at the same time" with a distributed system.
* A complementary argument to raising events before vs. after:  
   * Events represent things that have already happened in the past.  
   * If the state change that triggers an event is not committed, it's not externally observable.  
   * If you raise the events before the state change is committed, that potentially creates a race condition, because the consumer might read a stale state.
* An example where the the pair of actions was {persist to datastore1, persist to datastore2} as contrast to {persist to datastore, send email} would be provocative.  You would either end up with some nasty rollback/commit code, or need a distributed transaction.  Sometimes you can play with the order of operations to handle failure gracefully, but it's not a general solution.
* This could be another good place to re-iterate the value of DDD patterns and thinking, and why it has become more popular with microservices.  In this case, it guides the technical implementation to match the consistency rules of the business logic. (It's acceptable for the welcome email to be sent with some delay.  It could even be sent multiple times or even never in some edge cases, but those errors would be acceptable from a business perspective.  The worst case fallback might be a customer service ticket, which should be OK if it's rare.)
* Some discussion of idempotency and at-most-once, vs at-least-once delivery might be good.
* Some explicit discussion of CRUD vs CQRS might be good.  I'd guess that there will be a lot readers coming to this book from Django.